### PR TITLE
Use less stack space in the simplifier

### PR DESCRIPTION
**Pull request by abadams**
Get the large stack frames out of the recursive path.

Saves only ~20kb of stack on local laplacian, but saves 300kb on a huge unrolled summation. Not sure if it's worth merging, but opening it for discussion anyway.

Doesn't seem to affect compile times.

### CODE REVIEW

File: src/Simplify.cpp [hunk 1]
```diff
@@ -436,8 +436,27 @@ class Simplify : public IRMutator2 {
...
 
+    struct VisitResult {
+        Expr expr;
+        bool done;
+        template<typename T>
+        VisitResult(T &&t) : expr(std::forward<T>(t)), done(true) {}
+    };
+
+    template<typename T>
+    VisitResult remutate(T &&t) const {
```
Comment (steven-johnson) [inline src/Simplify.cpp]: I want to pronounced this "rem-you-tate"

**Comment by steven-johnson** [comment]
I like the stack savings, but am concerned that it makes a code pattern that is non-obvious and different from every other IRMutator in Halide. (Also: what's the downside if someone doesn't follow this pattern -- is it merely suboptimal or is there a risk on incorrectness?)

File: src/Simplify.cpp (hunk 1 above)
Comment (abadams) [reply src/Simplify.cpp]: "revisit" was another thing I considered.

**Comment by abadams** [comment]
Merely suboptimal - I didn't apply this change to all the mutators.

I'm trying something else to the simplifier to formalize the notion of a rewrite rule so that maybe we can test them more comprehensively. It may independently save stack space, so we'll see.

**Comment by abadams** [comment]
See #2845 as an alternative change that also reduces stack frame size

**Comment by abadams** [comment]
Closing in favor of #2845

**Pull request closed without merging**